### PR TITLE
#987 Use ElasticSearch Java Client API

### LINKED ISSUE #987: Use ElasticSearch Java Client API
**Describe the refactoring action**
We are currently using hand-crafted Java object and at times hand-crafted JSON objects to communicate with ElasticSearch servers. This makes our code unnecessarily complex. The benefit is basically saving some dependencies on ElasticSearch and thus potentially avoiding version conflicts e.g. on Lucene. However, we should investigate if we can use the ElasticSearch Java Client API without too much of a dependency impact to make our code more maintainable.

See: https://www.elastic.co/guide/en/elasticsearch/client/java-api/current/java-search.html

**Expected benefit**
Easier to read and more maintainable code, in particular as we make increased use of different ElasticSearch features and would piece-by-piece end up re-implementing the ElasticSearch Client API in our own codebase.

### PR DESCRIPTION
**Pull request by az79nefy**
**What's in the PR**
* Replace manual creation of ElasticSearch queries with Java Client API
* resolves #987 
- [x] restore functionality of randomizing the query results
- [x] fix JavaDoc compilation issue

**How to test manually**
* Execute external search queries

**Automatic testing**
* [X] PR includes unit tests

**Documentation**
* [ ] PR updates documentation

### CODE REVIEW

File: inception-external-search-elastic/src/main/java/de/tudarmstadt/ukp/inception/externalsearch/elastic/ElasticSearchProvider.java [hunk 1]
```diff
@@ -61,64 +63,69 @@
...
         String indexName = aTraits.getIndexName();
-        String searchPath = aTraits.getSearchPath();
+        String hostUrl = aTraits.getRemoteUrl().replaceFirst("https?://", "")
+                .replaceFirst("www.", "")
+                .split(":")[0];
         
-        // Set headers
-        HttpHeaders headers = new HttpHeaders();
-        headers.setContentType(MediaType.APPLICATION_JSON);
+        Settings settings = Settings.builder()
```
Comment (reckart) [inline inception-external-search-elastic/src/main/java/de/tudarmstadt/ukp/inception/externalsearch/elastic/ElasticSearchProvider.java]: What does this setting do?

File: inception-external-search-elastic/src/main/java/de/tudarmstadt/ukp/inception/externalsearch/elastic/ElasticSearchProvider.java [hunk 2]
```diff
@@ -61,64 +63,69 @@
...
+            client.addTransportAddress(
+                    new TransportAddress(
+                            InetAddress.getByName(hostUrl),
+                            9300));
         }
-        catch (JsonProcessingException e) {
-            log.error("Invalid JSON while building search query");
-            return Collections.emptyList();
+        catch (UnknownHostException e) {
+            System.out.println(e.getMessage());
```
Comment (reckart) [inline inception-external-search-elastic/src/main/java/de/tudarmstadt/ukp/inception/externalsearch/elastic/ElasticSearchProvider.java]: Please never print and exit ;) Probably best to throw the exception - or slightly worse to do what the previous code did: log and return no results.

File: inception-external-search-elastic/src/main/java/de/tudarmstadt/ukp/inception/externalsearch/elastic/ElasticSearchProvider.java [hunk 3]
```diff
@@ -61,64 +63,69 @@
...
+        SearchResponse response = client.prepareSearch(indexName)
+                .setQuery(QueryBuilders.termQuery(aTraits.getDefaultField(), aQuery))
+                .highlighter(highlightBuilder)
+                .setSize(aTraits.getResultSize())
+                .get();
 
-        for (ElasticSearchHit hit : queryResult.getHits().getHits()) {
-            if (hit.get_source() == null || hit.get_source().getMetadata() == null) {
+        for (SearchHit hit: response.getHits().getHits()) {
+            if (hit.getSourceAsMap() == null || hit.getSourceAsMap().get("metadata") == null) {
```
Comment (reckart) [inline inception-external-search-elastic/src/main/java/de/tudarmstadt/ukp/inception/externalsearch/elastic/ElasticSearchProvider.java]: Please introduce constants for strings such as "metadata". 

The need for the "metadata" and which fields this structure should provide should be added to our documentation where we explain what JSON structure we expect the documents in elastic search to have.

**Comment by reckart** [comment]
Jenkins, can you test this please?

File: inception-external-search-elastic/src/main/java/de/tudarmstadt/ukp/inception/externalsearch/elastic/ElasticSearchProvider.java [hunk 4]
```diff
@@ -61,64 +63,69 @@
...
+        HighlightBuilder.Field highlightField =
+                new HighlightBuilder.Field(aTraits.getDefaultField());
+        highlightField.highlighterType("unified");
+        highlightBuilder.field(highlightField);
 
-        // Send post query
-        queryResult = restTemplate.postForObject(searchUrl, entity,
-                ElasticSearchResult.class);
+        SearchResponse response = client.prepareSearch(indexName)
+                .setQuery(QueryBuilders.termQuery(aTraits.getDefaultField(), aQuery))
```
Comment (rgabbard) [inline inception-external-search-elastic/src/main/java/de/tudarmstadt/ukp/inception/externalsearch/elastic/ElasticSearchProvider.java]: It is not obvious to me that this code maintains the ability to do random ranking of query results?

**Comment by reckart** [comment]
@rgabbard thanks for checking in. We'll review if/how the randomness is preserved. ATM the build fails because of some odd JavaDoc problem.... once that is fixed, I'll have another look at the code.

**Comment by rgabbard** [comment]
@reckart : Thanks - we have an intern @qpwo who is looking into https://github.com/isi-vista/curated-training-annotator/issues/16 ; I'd be happy to have him test/fix the random ranking on top of this branch if it won't get in @az79nefy 's way.

**Comment by az79nefy** [comment]
Sorry, I have not finished this PR yet, it is still work in progress. I have not come around to do random ranking yet, but if @qpwo wants to look into it, I'd be happy for the help.

**Comment by reckart** [comment]
I have switched the code to use the REST high-level client which seems to be the API of choice these days for ElasticSearch.

I tried all kinds of tricks to fix the problem that ElasticSearch imports cannot be resolved by JavaDoc, but without success. When building using `mvn -Ddebug javadoc:aggregate`, a file `target/site/apidocs/options` is created which contains a list of all the JAR files passed to the JavaDoc compiler. While it contains many ElasticSearch JARs, `org.elasticsearch:elasticsearch` is not among them. This causes the imports not to resolve. I have tried:

- switching to older and newer versions of ElasticSearch
- changing the position of the ElasticSearch dependencies in the pom.xml file
- explicitly adding the type `jar` to the `org.elasticsearch:elasticsearch` dependency

... none worked. The RDF4J people hit the same problem though... maybe someone over there finds a solution: https://github.com/eclipse/rdf4j/issues/1148

File: inception-external-search-elastic/src/main/java/de/tudarmstadt/ukp/inception/externalsearch/elastic/ElasticSearchProvider.java (hunk 4 above)
Comment (az79nefy) [reply inception-external-search-elastic/src/main/java/de/tudarmstadt/ukp/inception/externalsearch/elastic/ElasticSearchProvider.java]: Random ranking of query results is now possible.

File: inception-external-search-elastic/src/main/java/de/tudarmstadt/ukp/inception/externalsearch/elastic/ElasticSearchProvider.java (hunk 3 above)
Comment (az79nefy) [reply inception-external-search-elastic/src/main/java/de/tudarmstadt/ukp/inception/externalsearch/elastic/ElasticSearchProvider.java]: Added constants for field names.
Modified documentation to further explain what fields are expected of the ElasticSearch documents.

**Comment by reckart** [comment]
For the time being, I am disabling the `javadoc:aggregate` target on Jenkins which trips over the ElasticSearch dependency - we don't ship the aggregate docs anyway. That should hopefully fix the build...

**Comment by reckart** [comment]
@az79nefy are there any code changes here that are left to be done?

**Comment by az79nefy** [comment]
Nope, no changes left to be done. This PR can be merged.

**Pull request merged**